### PR TITLE
make the switching station able to send supply and demand via digiline

### PR DESCRIPTION
**Pull request by DS-Minetest**
related: #168 
Give a switching station a mesecons signal and it will always send demand and supply (in a table) via digiline (with channel `"swtiching_station"..minetest.pos_to_string(pos)` where pos is the position of the switching station) when demand or supply changes.
Send via digiline "GET" or "get" and the switching station will send you on the same channel supply and demand back.

### CODE REVIEW

**Comment by DS-Minetest** [comment]
Still no comment. :confused:

File: technic/machines/switching_station.lua [hunk 1]
```diff
@@ -301,6 +318,20 @@ minetest.register_abm({
 				S("@1. Supply: @2 Demand: @3",
 				machine_name, technic.pretty_num(PR_eu_supply), technic.pretty_num(RE_eu_demand)))
 
+		-- digiline
+		if minetest.get_modpath("mesecons") and
+				minetest.get_modpath("digilines") and
```
Comment (numberZero) [inline technic/machines/switching_station.lua]: cache that

Comment (numberZero) [inline technic/machines/switching_station.lua]: cache that

File: technic/machines/switching_station.lua [hunk 2]
```diff
@@ -301,6 +318,20 @@ minetest.register_abm({
...
+				mesecon.is_powered(pos) then
+			if PR_eu_supply ~= meta:get_int("supply") or
+					RE_eu_demand ~= meta:get_int("demand") then
+				digilines.receptor_send(pos, digilines.rules.default,
+						"switching_station"..minetest.pos_to_string(pos), {
```
Comment (numberZero) [inline technic/machines/switching_station.lua]: channel should be configurable

File: technic/machines/switching_station.lua [hunk 3]
```diff
@@ -63,6 +63,23 @@ minetest.register_node("technic:switching_station",{
...
 		minetest.forceload_free_block(pos)
 	end,
+	mesecons = {effector = {
+		rules = mesecon.rules.default,
+	}},
+	digiline = {
+		receptor = {action = function() end},
+		effector = {
+			action = function(pos, node, channel, msg)
+				if msg == "GET" or msg == "get" then
```
Comment (numberZero) [inline technic/machines/switching_station.lua]: no check for channel. imagine you have several switching stations (e.g. MV and HV) connected to a single digiline wire.

**Comment by MT-Modder** [comment]
I honestly can't wait! You've tested it (extensively)?

File: technic/machines/switching_station.lua [hunk 4]
```diff
@@ -301,6 +318,20 @@ minetest.register_abm({
...
 
+		-- digiline
+		if minetest.get_modpath("mesecons") and
+				minetest.get_modpath("digilines") and
+				mesecon.is_powered(pos) then
+			if PR_eu_supply ~= meta:get_int("supply") or
+					RE_eu_demand ~= meta:get_int("demand") then
+				digilines.receptor_send(pos, digilines.rules.default,
+						"switching_station"..minetest.pos_to_string(pos), {
+							supply = PR_eu_supply,
```
Comment (numberZero) [inline technic/machines/switching_station.lua]: 7th indentation level? is that really necessary?

File: technic/machines/switching_station.lua (hunk 3 above)
Comment (DS-Minetest) [inline technic/machines/switching_station.lua]: So again use `"swtiching_station"..minetest.pos_to_string(pos)` as channel?

**Comment by DS-Minetest** [comment]
@MT-Modder How often will you write this?

File: technic/machines/switching_station.lua (hunk 4 above)
Comment (DS-Minetest) [inline technic/machines/switching_station.lua]: I think so.

**Comment by DS-Minetest** [comment]
I need to do more testing.

**Comment by numberZero** [comment]
er, I meant to cache the boolean value, e.g.
`local use_digilines = minetest.get_modpath("mesecons") and minetest.get_modpath("digilines") and true`
although this way it is readable too, so not a problem.
(btw, as you use soft-dependency, you can use smth like `minetest.global_exists('mesecon')` instead of `get_modpath`. but I am not sure what is better)

**Comment by numberZero** [comment]
so... waiting for this to be merged =) that would be a great thing for automation)
@DS-Minetest I wanted to ask you to do the same for battery boxes (as another PR, of course) but noticed you made this PR from `master`. Don’t do that next time, at least because it prevents you from making several PRs at once. Use feature branches. (ex.: `git checkout -b digiline-switching-station`)

**Comment by DS-Minetest** [comment]
Yes, I'm very unexperienced with git, I always do it with master and I should look some time how to do it better.

Perhaps I should check instead for the global `digilines` and `mesecons` variables.

Note, this is still nearly not tested, I don't even know, if it shuts down if mesecons or digilines is not installed or if different channels work.

**Comment by kaeza** [comment]
It seems the channel is not being saved?

I don't have much experience with actual use of digilines, so can somebody show an example setup for testing?

**Comment by numberZero** [comment]
oh well, it needs to handle `on_receive_fields`...

**Comment by kaeza** [comment]
@numberZero Thanks! It seems to work. Will add the `on_receive_fields` myself and merge it.

**Comment by kaeza** [comment]
https://github.com/minetest-technic/technic/commit/6abd857b3fc0dab74c71d77a26b05e412d43f562

Thank you for the contribution!

**Comment by DS-Minetest** [comment]
> it needs to handle `on_receive_fields`

Yep, that was the thing that was missing, I would have added it after testing.

This was a very fast merge, I hope there are no bugs.

> Perhaps I should check instead for the global digilines and mesecons variables.

I still don't know, would it be better than checking for modpath? I mean checking for a variable should be lighter than using a function for search for the modpath.

Todo:
> do the same for battery boxes

**Comment by DS-Minetest** [comment]
_If anyone wants to know (probably not),_ I've started to make something similar to battery boxes: https://github.com/DS-Minetest/technic/tree/battery_box-digilines
A PR will be started when I've finished it.

@numberZero thanks for the information about the `git checkout -b` thing!

**Comment by kaeza** [comment]
>I still don't know, would it be better than checking for modpath? I mean checking for a variable should be lighter than using a function for search for the modpath.

That is run at load time, so doesn't matter.

>Yep, that was the thing that was missing, I would have added it after testing.
>
>This was a very fast merge, I hope there are no bugs.

I tested with #0's code and it worked as I would expect. If somebody discovers bugs, they can be fixed. It's not like the codebase is now magically locked :)

It's fine to do devices in separate commits. Makes it easier to revert individual changes if something goes wrong.

**Pull request closed without merging**